pkg/webhook/certs.go

```diff
@@ -18,25 +18,74 @@ package webhook
 
 import (
 	"crypto/tls"
+	"github.com/golang/glog"
 	"io/ioutil"
+	"time"
 )
 
-// certBundle is a container of a X509 certificate file and a corresponding key file for the
+// certProvider is a container of a X509 certificate file and a corresponding key file for the
 // webhook server, and a CA certificate file for the API server to verify the server certificate.
-type certBundle struct {
+type certProvider struct {
 	serverCertFile string
 	serverKeyFile  string
 	caCertFile     string
+	reloadInterval time.Duration
+	ticker         *time.Ticker
+	stopChannel    *chan interface{}
+	currentCert    *tls.Certificate
 }
 
-// configServerTLS configures TLS for the admission webhook server.
-func configServerTLS(certBundle *certBundle) (*tls.Config, error) {
-	cert, err := tls.LoadX509KeyPair(certBundle.serverCertFile, certBundle.serverKeyFile)
+func NewCertProvider(serverCertFile, serverKeyFile, caCertFile string, reloadInterval time.Duration) (*certProvider, error) {
+	cert, err := tls.LoadX509KeyPair(serverCertFile, serverKeyFile)
 	if err != nil {
 		return nil, err
 	}
+	return &certProvider{
+		serverCertFile: serverCertFile,
+		serverKeyFile:  serverKeyFile,
+		caCertFile:     caCertFile,
+		reloadInterval: reloadInterval,
+		currentCert:    &cert,
+	}, nil
+}
+
+func (c *certProvider) Start() {
+	stopChannel := make(chan interface{})
+	c.stopChannel = &stopChannel
+	ticker := time.NewTicker(c.reloadInterval)
+	c.ticker = ticker
+	go func() {
+		for {
+			select {
+			case <-stopChannel:
+				return
+			case <-c.ticker.C:
+				c.updateCert()
+			}
+		}
+	}()
+}
+
+func (c *certProvider) tlsConfig() *tls.Config {
+	return &tls.Config{
+		GetCertificate: func(ch *tls.ClientHelloInfo) (*tls.Certificate, error) {
+			return c.currentCert, nil
+		},
+	}
+}
+
+func (c *certProvider) Stop() {
+	close(*c.stopChannel)
+	c.ticker.Stop()
+}
 
-	return &tls.Config{Certificates: []tls.Certificate{cert}}, nil
+func (c *certProvider) updateCert() {
+	cert, err := tls.LoadX509KeyPair(c.serverCertFile, c.serverKeyFile)
+	if err != nil {
+		glog.Error("Could not reload certificate: ", err)
+		return
+	}
+	c.currentCert = &cert
```

Comment:
This only updates `currentCert`. What about making the new certificate effective? I think two things need to be done: 1) update the `MutatingWebhookConfigurations ` with the new cert, and 2) restart the web server to load the new cert.

Reply:
`GetCertificate` in `tls.Config` is a callback used in the TLS handshake step to get the current certificate. Our implementation always returns `currentCert`, so this is all that's needed to make the new cert effective. See:

https://golang.org/pkg/crypto/tls/#Config
https://github.com/GoogleCloudPlatform/spark-on-k8s-operator/blob/9c1e54b23f12565e8d10704b4991695951dc1ed4/pkg/webhook/certs.go#L71-L73

Reply:
Ah, got it. So this handshake also registers the new cert with the API server?

Reply:
The API server doesn't need to be notified of new server certs as the cert chain should roll up to the configured CA cert.

Reply:
I see. Is this PR tested in production?

Reply:
Yes, we've been running this in prod for a few months.

Reply:
Cool!